PLFM 2114, 2124

kaysoky opened the pull request:
- Changes DBOUserGroup to include an extra ETAG column.
- GroupMembers DAO Declaration
- Crowd GroupMembers DAO implementation
- DBO GroupMembers DAO implementation
- Tests (with migration test)
- Worker for migrating Crowd groups to RDS groups

kaysoky committed:
DBO for GroupMembers table

+ Changes DBOUserGroup to include an extra ETAG column.

kaysoky committed:
GroupMembers DAO

+ Declaration
+ Crowd implementation
+ DBO implementation
+ Tests (with migration test)

kaysoky committed:
Periodic Crowd to RDS worker

The implementation is sort-of brute-force (which might be ok since there aren't that many groups/memberships in Prod).
- The "updatedDate" for Crowd Users/Groups is not affected by membership.
- Crowd Groups do not have any properties/attributes by default, other than a name.

kaysoky committed:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/Synapse-Repository-Services into PLFM-2114

kaysoky commented:
These changes are mostly additive and should not break existing tests.  
The possibly breaking change of the UserManager (https://sagebionetworks.jira.com/browse/PLFM-2128) will be included in a separate pull once this one is reviewed/pulled.

kaysoky commented:
https://github.com/kaysoky/Synapse-Repository-Services/compare/PLFM-2114...PLFM-2128

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java:
```diff
@@ -0,0 +1,151 @@
+			}
+			
+			// Update the etags of all affected items
+			List<String> ids = new ArrayList<String>();
```
this dao cannot touch the DB

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java:
```diff
@@ -0,0 +1,151 @@
+		try {
+			group = userGroupDAO.get(principalId);
+		} catch (NotFoundException e) {
+			throw new DatastoreException(e);
```
this should throw a not found

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java:
```diff
@@ -0,0 +1,151 @@
+	@Override
+	public void addMembers(GroupMembers dto) throws DatastoreException {
+		// Make sure the DTO holds a group, not an individual
+		UserGroup group;
```
throw exception if more than on as there is no transaction.

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java:
```diff
@@ -0,0 +1,151 @@
+			for (UserGroup removed : dto.getMembers()) {
+				ids.add(removed.getId());
+			}
+			userGroupDAO.touchList(ids);
```
again no db touch here.

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java:
```diff
@@ -0,0 +1,151 @@
+	@Override
+	public void removeMembers(GroupMembers dto) throws DatastoreException {
+		UserGroup group;
+		try {
```
only one as well

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java:
```diff
@@ -0,0 +1,151 @@
+		try {
+			group = userGroupDAO.get(dto.getId());
+		} catch (NotFoundException e) {
+			throw new DatastoreException(e);
```
not found

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java:
```diff
@@ -0,0 +1,151 @@
+
+	@Override
+	public void removeMembers(GroupMembers dto) throws DatastoreException {
+		UserGroup group;
```
removeMember(string groupId, List<String> listOfIdsToRemove)

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java:
```diff
@@ -0,0 +1,151 @@
+	@Transactional(readOnly = false, propagation = Propagation.REQUIRED)
+	@Override
+	public void addMembers(GroupMembers dto) throws DatastoreException {
+		// Make sure the DTO holds a group, not an individual
```
same signature as remove()

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java:
```diff
@@ -0,0 +1,151 @@
+	
+	@Override
+	public GroupMembers getMembers(String principalId)
+			throws DatastoreException {
```
return List<UserGroup>

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOGroupMembersDAOImpl.java:
```diff
@@ -0,0 +1,186 @@
+			" WHERE "+SqlConstants.COL_USER_GROUP_ID+" IN"+
+					" (SELECT "+SqlConstants.COL_GROUP_MEMBERS_MEMBER_ID+" FROM "+SqlConstants.TABLE_GROUP_MEMBERS+
+					" WHERE "+SqlConstants.COL_GROUP_MEMBERS_GROUP_ID+"=:"+PRINCIPAL_ID_PARAM_NAME+")";
+	
```
re-write as a single non-nested

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOGroupMembersDAOImpl.java:
```diff
@@ -0,0 +1,186 @@
+			UserGroupUtils.copyDboToDto(dbo, dto);
+			members.getMembers().add(dto);
+		}
+
```
update the util to take a list (any other place that would use that should be refactored)

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOUserGroupDAOImpl.java:
```diff
@@ -359,4 +398,27 @@ public void bootstrapUsers() throws Exception {
+	@Transactional(readOnly = false, propagation = Propagation.REQUIRED)
+	@Override
+	public void touchList(List<String> ids) {
+		MapSqlParameterSource params[] = new MapSqlParameterSource[ids.size()];
```
needs deadlock prevention

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOGroupMembersDAOImpl.java:
```diff
@@ -0,0 +1,186 @@
+	private Set<String> getParentGroupIds(String principalId) {
+		Set<String> groupIds = new HashSet<String>();
+		List<String> memberIds = new ArrayList<String>();
+		memberIds.add(principalId);
```
rename explored and frontier

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOUserGroupDAOImpl.java:
```diff
@@ -359,4 +398,27 @@ public void bootstrapUsers() throws Exception {
+	}
+
+	@Override
+	public boolean hasChanged(UserGroup dto) {
```
remove if can

john-hill commented on services/workers/src/main/resources/grouper-spb.xml:
```diff
@@ -0,0 +1,43 @@
+		<!-- We stagger the start delay of each trigger to spread out the timing -->
+		<property name="startDelay" value="567" />
+		<property name="repeatInterval" value="300000" />
+	</bean>
```
make sure this is the same as the current cache time.

kaysoky committed:
Based on code review:

- GroupMembers DTO has been removed
- DAO for GroupMembers now takes primarily strings as arguments
- Extra DBO made for caching the parents of UserGroups
+ Deadlock consideration

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/CrowdGroupMembersDAO.java:
```diff
@@ -0,0 +1,152 @@
+	}
+
+	@Override
+	public Set<String> markAsUpdated(String groupId, List<String> memberIds)
```
try to make private.

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOGroupMembersDAOImpl.java:
```diff
@@ -0,0 +1,319 @@
+		markAsUpdated(groupId, memberIds);
+	}
+	
+	public static byte[] zipParents(List<String> parents) throws IOException {
```
move to utility

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOGroupMembersDAOImpl.java:
```diff
@@ -0,0 +1,319 @@
+	}
+
+	@Override
+	public Set<String> markAsUpdated(String groupId, List<String> memberIds)
```
multi-thread test.

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOUserGroupDAOImpl.java:
```diff
@@ -82,15 +86,20 @@
-		" p ON g."+COL_USER_GROUP_ID+" = p."+COL_USER_PROFILE_ID+
-		" ORDER BY g."+COL_USER_GROUP_ID+" LIMIT :"+LIMIT_PARAM_NAME+
-		" OFFSET :"+OFFSET_PARAM_NAME;
+	// private static final String SELECT_ALL_PAGINATED_WITH_ETAG = 
```
remove
